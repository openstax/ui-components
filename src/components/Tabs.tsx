import React from "react";
import * as AC from "react-aria-components";
import { colors } from "../theme";
import styled, { css } from "styled-components";

type TabsProps = {
  variant?: "button-bar";
  size?: "large" | "small";
  className?: string;
  children?: React.ReactNode;
} & AC.TabsProps;

const buttonBarCss = css`
  .react-aria-TabList {
    border: 0.1rem solid ${colors.palette.pale};
    border-radius: 0.5rem;
  }

  .react-aria-Tab {
    padding: 0 1.6rem;
    min-height: ${({ size }: TabsProps) =>
      size === "large" ? "4.8rem" : "2.8rem"};

    &[data-selected] {
      background: ${colors.palette.neutralLight};
    }
    &:hover:not([data-selected]) {
      background: ${colors.palette.neutralLighter};
    }
  }

  &[data-orientation="horizontal"] .react-aria-Tab {
    border-right: 0.1rem solid ${colors.palette.pale};

    &:first-child {
      border-top-left-radius: 0.4rem;
      border-bottom-left-radius: 0.4rem;
      border-left: 0;
    }
    &:last-child {
      border-top-right-radius: 0.4rem;
      border-bottom-right-radius: 0.4rem;
      border-right: 0;
    }
  }

  &[data-orientation="vertical"] .react-aria-Tab {
    border-bottom: 0.1rem solid ${colors.palette.pale};

    &:first-child {
      border-top-left-radius: 0.4rem;
      border-top-right-radius: 0.4rem;
      border-top: 0;
    }
    &:last-child {
      border-bottom-left-radius: 0.4rem;
      border-bottom-right-radius: 0.4rem;
      border-bottom: 0;
    }
  }
`;

const tabsCss = css`
  .react-aria-TabList {
    border-bottom: 0.1rem solid ${colors.palette.pale};
  }

  &[data-orientation="vertical"] .react-aria-TabList {
    border: none;
  }

  .react-aria-Tab {
    padding: 0 1.6rem;
    height: ${({ size }: TabsProps) =>
      size === "large" ? "4.8rem" : "2.8rem"};
    border-bottom: 0.4rem solid transparent;

    &[data-selected] {
      border-color: #63a524;
    }
    &:hover:not([data-selected]) {
      border-color: ${colors.palette.neutralLighter};
    }
  }
`;

const StyledTabs = styled(AC.Tabs)`
  .react-aria-TabList {
    overflow-y: scroll;
    overscroll-behavior: contain;
    display: flex;
  }

  &[data-orientation="vertical"] .react-aria-TabList {
    flex-direction: column;
    overflow: scroll;
  }
  &[data-orientation="horizontal"] .react-aria-TabList {
    flex-direction: row;
  }

  .react-aria-Tab {
    flex: 1 1 auto;
    display: flex;
    align-items: center;
    justify-content: center;
    outline-offset: -0.1rem; // Prevent overflow scroll from clipping outline
    white-space: nowrap;
    font-size: ${({ size }: TabsProps) =>
      size === "large" ? "1.6rem" : "1.4rem"};

    &:hover {
      cursor: pointer;
    }
  }

  ${(props: TabsProps) =>
    props.variant === "button-bar" ? buttonBarCss : tabsCss}
`;

export const Tabs = ({
  variant,
  size = "large",
  className,
  children,
  ...restProps
}: TabsProps & AC.TabsProps) => {
  return (
    <StyledTabs
      variant={variant}
      size={size}
      className={className}
      {...restProps}
    >
      {children}
    </StyledTabs>
  );
};

export { TabList, Tab, TabPanel } from "react-aria-components";
