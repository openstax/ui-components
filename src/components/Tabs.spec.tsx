import { render } from "@testing-library/react";
import { Tabs, Tab, TabList, TabPanel } from "./Tabs";

describe("Tabs component", () => {
  describe("normal styling", () => {
    it("renders correctly", () => {
      const { asFragment } = render(
        <Tabs>
          <TabList aria-label="Items">
            <Tab id="one">First Item</Tab>
            <Tab id="two">Second Item</Tab>
            <Tab id="three">Last Item</Tab>
          </TabList>
          <TabPanel id="one">First Content Panel</TabPanel>
          <TabPanel id="two">Second Content Panel</TabPanel>
          <TabPanel id="three">Third Content Panel</TabPanel>
        </Tabs>,
      );
      expect(asFragment()).toMatchSnapshot();
    });

    it("renders correctly with small size", () => {
      const { asFragment } = render(
        <Tabs size="small">
          <TabList aria-label="Items">
            <Tab id="one">First Item</Tab>
            <Tab id="two">Second Item</Tab>
            <Tab id="three">Last Item</Tab>
          </TabList>
          <TabPanel id="one">First Content Panel</TabPanel>
          <TabPanel id="two">Second Content Panel</TabPanel>
          <TabPanel id="three">Third Content Panel</TabPanel>
        </Tabs>,
      );
      expect(asFragment()).toMatchSnapshot();
    });

    it("defaults size", () => {
      const { asFragment } = render(
        <Tabs size={undefined}>
          <TabList aria-label="Items">
            <Tab id="one">First Item</Tab>
            <Tab id="two">Second Item</Tab>
            <Tab id="three">Last Item</Tab>
          </TabList>
          <TabPanel id="one">First Content Panel</TabPanel>
          <TabPanel id="two">Second Content Panel</TabPanel>
          <TabPanel id="three">Third Content Panel</TabPanel>
        </Tabs>,
      );
      expect(asFragment()).toMatchSnapshot();
    });
  });

  describe("button-bar styling", () => {
    it("renders correctly with button-bar variant", () => {
      const { asFragment } = render(
        <Tabs variant="button-bar">
          <TabList aria-label="Items">
            <Tab id="one">First Item</Tab>
            <Tab id="two">Second Item</Tab>
            <Tab id="three">Last Item</Tab>
          </TabList>
          <TabPanel id="one">First Content Panel</TabPanel>
          <TabPanel id="two">Second Content Panel</TabPanel>
          <TabPanel id="three">Third Content Panel</TabPanel>
        </Tabs>,
      );
      expect(asFragment()).toMatchSnapshot();
    });

    it("renders correctly with button-bar variant and small size", () => {
      const { asFragment } = render(
        <Tabs size="small" variant="button-bar">
          <TabList aria-label="Items">
            <Tab id="one">First Item</Tab>
            <Tab id="two">Second Item</Tab>
            <Tab id="three">Last Item</Tab>
          </TabList>
          <TabPanel id="one">First Content Panel</TabPanel>
          <TabPanel id="two">Second Content Panel</TabPanel>
          <TabPanel id="three">Third Content Panel</TabPanel>
        </Tabs>,
      );
      expect(asFragment()).toMatchSnapshot();
    });

    it("defaults size", () => {
      const { asFragment } = render(
        <Tabs size={undefined}>
          <TabList aria-label="Items">
            <Tab id="one">First Item</Tab>
            <Tab id="two">Second Item</Tab>
            <Tab id="three">Last Item</Tab>
          </TabList>
          <TabPanel id="one">First Content Panel</TabPanel>
          <TabPanel id="two">Second Content Panel</TabPanel>
          <TabPanel id="three">Third Content Panel</TabPanel>
        </Tabs>,
      );
      expect(asFragment()).toMatchSnapshot();
    });
  });
});
