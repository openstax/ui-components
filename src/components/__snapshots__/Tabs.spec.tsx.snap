// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tabs component button-bar styling renders 1`] = `
<DocumentFragment>
  <div
    class="sc-bczRLJ kWkDUg"
    data-orientation="horizontal"
    data-rac=""
  >
    <div
      aria-label="Items"
      aria-orientation="horizontal"
      class="react-aria-TabList"
      data-orientation="horizontal"
      data-rac=""
      id="react-aria-13"
      role="tablist"
    >
      <div
        aria-controls="react-aria-13-tabpanel-one"
        aria-selected="true"
        class="react-aria-Tab"
        data-key="one"
        data-rac=""
        data-selected="true"
        id="react-aria-13-tab-one"
        role="tab"
        tabindex="0"
      >
        First Item
      </div>
      <div
        aria-selected="false"
        class="react-aria-Tab"
        data-key="two"
        data-rac=""
        id="react-aria-13-tab-two"
        role="tab"
        tabindex="-1"
      >
        Second Item
      </div>
      <div
        aria-selected="false"
        class="react-aria-Tab"
        data-key="three"
        data-rac=""
        id="react-aria-13-tab-three"
        role="tab"
        tabindex="-1"
      >
        Last Item
      </div>
    </div>
    <div
      aria-labelledby="react-aria-13-tab-one"
      class="react-aria-TabPanel"
      data-rac=""
      id="react-aria-13-tabpanel-one"
      role="tabpanel"
      tabindex="0"
    >
      First Content Panel
    </div>
  </div>
</DocumentFragment>
`;

exports[`Tabs component button-bar styling renders large size 1`] = `
<DocumentFragment>
  <div
    class="sc-bczRLJ kTnZrf"
    data-orientation="horizontal"
    data-rac=""
  >
    <div
      aria-label="Items"
      aria-orientation="horizontal"
      class="react-aria-TabList"
      data-orientation="horizontal"
      data-rac=""
      id="react-aria-21"
      role="tablist"
    >
      <div
        aria-controls="react-aria-21-tabpanel-one"
        aria-selected="true"
        class="react-aria-Tab"
        data-key="one"
        data-rac=""
        data-selected="true"
        id="react-aria-21-tab-one"
        role="tab"
        tabindex="0"
      >
        First Item
      </div>
      <div
        aria-selected="false"
        class="react-aria-Tab"
        data-key="two"
        data-rac=""
        id="react-aria-21-tab-two"
        role="tab"
        tabindex="-1"
      >
        Second Item
      </div>
      <div
        aria-selected="false"
        class="react-aria-Tab"
        data-key="three"
        data-rac=""
        id="react-aria-21-tab-three"
        role="tab"
        tabindex="-1"
      >
        Last Item
      </div>
    </div>
    <div
      aria-labelledby="react-aria-21-tab-one"
      class="react-aria-TabPanel"
      data-rac=""
      id="react-aria-21-tabpanel-one"
      role="tabpanel"
      tabindex="0"
    >
      First Content Panel
    </div>
  </div>
</DocumentFragment>
`;

exports[`Tabs component button-bar styling renders small size 1`] = `
<DocumentFragment>
  <div
    class="sc-bczRLJ kFnuyo"
    data-orientation="horizontal"
    data-rac=""
  >
    <div
      aria-label="Items"
      aria-orientation="horizontal"
      class="react-aria-TabList"
      data-orientation="horizontal"
      data-rac=""
      id="react-aria-17"
      role="tablist"
    >
      <div
        aria-controls="react-aria-17-tabpanel-one"
        aria-selected="true"
        class="react-aria-Tab"
        data-key="one"
        data-rac=""
        data-selected="true"
        id="react-aria-17-tab-one"
        role="tab"
        tabindex="0"
      >
        First Item
      </div>
      <div
        aria-selected="false"
        class="react-aria-Tab"
        data-key="two"
        data-rac=""
        id="react-aria-17-tab-two"
        role="tab"
        tabindex="-1"
      >
        Second Item
      </div>
      <div
        aria-selected="false"
        class="react-aria-Tab"
        data-key="three"
        data-rac=""
        id="react-aria-17-tab-three"
        role="tab"
        tabindex="-1"
      >
        Last Item
      </div>
    </div>
    <div
      aria-labelledby="react-aria-17-tab-one"
      class="react-aria-TabPanel"
      data-rac=""
      id="react-aria-17-tabpanel-one"
      role="tabpanel"
      tabindex="0"
    >
      First Content Panel
    </div>
  </div>
</DocumentFragment>
`;

exports[`Tabs component normal styling renders 1`] = `
<DocumentFragment>
  <div
    class="sc-bczRLJ lljHPQ"
    data-orientation="horizontal"
    data-rac=""
  >
    <div
      aria-label="Items"
      aria-orientation="horizontal"
      class="react-aria-TabList"
      data-orientation="horizontal"
      data-rac=""
      id="react-aria-1"
      role="tablist"
    >
      <div
        aria-controls="react-aria-1-tabpanel-one"
        aria-selected="true"
        class="react-aria-Tab"
        data-key="one"
        data-rac=""
        data-selected="true"
        id="react-aria-1-tab-one"
        role="tab"
        tabindex="0"
      >
        First Item
      </div>
      <div
        aria-selected="false"
        class="react-aria-Tab"
        data-key="two"
        data-rac=""
        id="react-aria-1-tab-two"
        role="tab"
        tabindex="-1"
      >
        Second Item
      </div>
      <div
        aria-selected="false"
        class="react-aria-Tab"
        data-key="three"
        data-rac=""
        id="react-aria-1-tab-three"
        role="tab"
        tabindex="-1"
      >
        Last Item
      </div>
    </div>
    <div
      aria-labelledby="react-aria-1-tab-one"
      class="react-aria-TabPanel"
      data-rac=""
      id="react-aria-1-tabpanel-one"
      role="tabpanel"
      tabindex="0"
    >
      First Content Panel
    </div>
  </div>
</DocumentFragment>
`;

exports[`Tabs component normal styling renders large size 1`] = `
<DocumentFragment>
  <div
    class="sc-bczRLJ AsoTD"
    data-orientation="horizontal"
    data-rac=""
  >
    <div
      aria-label="Items"
      aria-orientation="horizontal"
      class="react-aria-TabList"
      data-orientation="horizontal"
      data-rac=""
      id="react-aria-9"
      role="tablist"
    >
      <div
        aria-controls="react-aria-9-tabpanel-one"
        aria-selected="true"
        class="react-aria-Tab"
        data-key="one"
        data-rac=""
        data-selected="true"
        id="react-aria-9-tab-one"
        role="tab"
        tabindex="0"
      >
        First Item
      </div>
      <div
        aria-selected="false"
        class="react-aria-Tab"
        data-key="two"
        data-rac=""
        id="react-aria-9-tab-two"
        role="tab"
        tabindex="-1"
      >
        Second Item
      </div>
      <div
        aria-selected="false"
        class="react-aria-Tab"
        data-key="three"
        data-rac=""
        id="react-aria-9-tab-three"
        role="tab"
        tabindex="-1"
      >
        Last Item
      </div>
    </div>
    <div
      aria-labelledby="react-aria-9-tab-one"
      class="react-aria-TabPanel"
      data-rac=""
      id="react-aria-9-tabpanel-one"
      role="tabpanel"
      tabindex="0"
    >
      First Content Panel
    </div>
  </div>
</DocumentFragment>
`;

exports[`Tabs component normal styling renders small size 1`] = `
<DocumentFragment>
  <div
    class="sc-bczRLJ mEdVR"
    data-orientation="horizontal"
    data-rac=""
  >
    <div
      aria-label="Items"
      aria-orientation="horizontal"
      class="react-aria-TabList"
      data-orientation="horizontal"
      data-rac=""
      id="react-aria-5"
      role="tablist"
    >
      <div
        aria-controls="react-aria-5-tabpanel-one"
        aria-selected="true"
        class="react-aria-Tab"
        data-key="one"
        data-rac=""
        data-selected="true"
        id="react-aria-5-tab-one"
        role="tab"
        tabindex="0"
      >
        First Item
      </div>
      <div
        aria-selected="false"
        class="react-aria-Tab"
        data-key="two"
        data-rac=""
        id="react-aria-5-tab-two"
        role="tab"
        tabindex="-1"
      >
        Second Item
      </div>
      <div
        aria-selected="false"
        class="react-aria-Tab"
        data-key="three"
        data-rac=""
        id="react-aria-5-tab-three"
        role="tab"
        tabindex="-1"
      >
        Last Item
      </div>
    </div>
    <div
      aria-labelledby="react-aria-5-tab-one"
      class="react-aria-TabPanel"
      data-rac=""
      id="react-aria-5-tabpanel-one"
      role="tabpanel"
      tabindex="0"
    >
      First Content Panel
    </div>
  </div>
</DocumentFragment>
`;
